[question] visible=True for tool requires seems not to work correctly

### What is your question?

Hi!
I have a package A that is dependent on package B. Both must use same version of tool X and I set the visible attribute to True.
recipe for B:
```
import os
import sys
from conan import ConanFile
from conan.tools.files import copy, save
from conan.tools.cmake import CMake, CMakeToolchain, CMakeDeps


class pkgb(ConanFile):
    name = 'pkgb'
    url = 'https://github.com/pkgb.git'
    exports_sources = "src/*", "archi/*", "doc/*", "inc/*", "CMakeLists.txt", "conf/*"
    settings = "os", "compiler", "arch", "build_type"
    python_requires = "common/2.0.0@autosar/release"
    python_requires_extend = "common.AutosarBase"

    def build_requirements(self):
        self.tool_requires("ctool/3.0.0@autosar/release", visible=True)
        
    def build(self):
        # we can skip build and skip tests
        # this allows us to run conan build for e.g. executing single tests
        skip_build = self.conf.get("user.build:skip_build", default=False)

        cmake = CMake(self)
        
        if skip_build:
            print(f"Build skipped: {skip_build}")
            cmake.configure(variables={'PYTHON_EXE': sys.executable, 'BUILD_TESTING': 'ON'} )
        else:
            cmake.configure(variables={'PYTHON_EXE': sys.executable})
            cmake.build()
        
        if not self.conf.get("tools.build:skip_test", default=False):
            print("Tests are set to:", self.conf.get("tools.build:skip_test"))
            ctests = self.conf.get("user.build:ctests", default="")
            verbose = '--verbose' if self.conf.get("user.build:verbose", default=False) else ''
            self.run(f"ctest {verbose} --output-on-failure -R \"{ctests}\"")

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()
        deps = CMakeDeps(self)
        deps.generate()

    def package_info(self):
        self.cpp_info.includedirs = ['inc']
        self.cpp_info.libdirs = ['lib']
        self.cpp_info.objects = [os.path.join("lib", "packageb.c.obj")]
# EOF
```
Example for recipe a:
```
import os
import sys
from conan import ConanFile
from conan.tools.files import copy, save
from conan.tools.cmake import CMake, CMakeToolchain, CMakeDeps


class pkga(ConanFile):
    name = 'pkga'
    url = 'https://github.com/pkga.git'
    exports_sources = "src/*", "archi/*", "doc/*", "inc/*", "CMakeLists.txt", "conf/*"
    settings = "os", "compiler", "arch", "build_type"
    python_requires = "common/2.0.0@autosar/release"
    python_requires_extend = "common.AutosarBase"

	
    def requirements(self):
        self.requires("pkgb/1.0.0@autosar/release")
	
    def build_requirements(self):
        self.tool_requires("ctool/3.0.0@autosar/release", visible=True)
        
    def build(self):
        # we can skip build and skip tests
        # this allows us to run conan build for e.g. executing single tests
        skip_build = self.conf.get("user.build:skip_build", default=False)

        cmake = CMake(self)
        
        if skip_build:
            print(f"Build skipped: {skip_build}")
            cmake.configure(variables={'PYTHON_EXE': sys.executable, 'BUILD_TESTING': 'ON'} )
        else:
            cmake.configure(variables={'PYTHON_EXE': sys.executable})
            cmake.build()
        
        if not self.conf.get("tools.build:skip_test", default=False):
            print("Tests are set to:", self.conf.get("tools.build:skip_test"))
            ctests = self.conf.get("user.build:ctests", default="")
            verbose = '--verbose' if self.conf.get("user.build:verbose", default=False) else ''
            self.run(f"ctest {verbose} --output-on-failure -R \"{ctests}\"")

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()
        deps = CMakeDeps(self)
        deps.generate()

    def package_info(self):
        self.cpp_info.includedirs = ['inc']
        self.cpp_info.libdirs = ['lib']
        self.cpp_info.objects = [os.path.join("lib", "packageb.c.obj")]
# EOF
```

CMakeLists for pkga has: 
`find_package(b config REQUIRED)`
pkg a fails with :
-- Conan: Target declared 'pkgb::pkgb'
CMake Error at C:/ProgramFiles/CMake/share/cmake-3.26/Modules/CMakeFindDependencyMacro.cmake:76 (find_package):
  Could not find a package configuration file provided by "ctool" with any
  of the following names:

    ctoolConfig.cmake
    ctool-config.cmake
  Add the installation prefix of "ctool" to CMAKE_PREFIX_PATH or set
  "ctool_DIR" to a directory containing one of the above files.  If "ctool"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  kdp/build/Release/generators/pkgb-config.cmake:24 (find_dependency)
  _src/CMakeLists.txt:6 (find_package)
What I donot understand why is it looking for ctool in find_package (b CONFIG REQUIRED)
Am I not using the visible attribute in the correct way?

Could you please help me?

Thanks!
Br,
Maitrey

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

--- a/conans/model/dependencies.py
+++ b/conans/model/dependencies.py
@@ -165,7 +165,9 @@ def get_transitive_requires(consumer, dependency):
     """ the transitive requires that we need are the consumer ones, not the current dependencey
     ones, so we get the current ones, then look for them in the consumer, and return those
     """
-    pkg_deps = dependency.dependencies.filter({"direct": True})
+    # The build dependencies cannot be transitive in generators like CMakeDeps,
+    # even if users make them visible
+    pkg_deps = dependency.dependencies.filter({"direct": True, "build": False})
     result = consumer.dependencies.transitive_requires(pkg_deps)
     result = result.filter({"skip": False})
     return result

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_build_context_transitive_build.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_build_context_transitive_build.py
@@ -137,19 +137,23 @@ def test_error_cmakedeps_transitive_build_requires():
     assert "tool/0.1: Created package" in c.out
 
 
-def test():
+def test_transitive_tool_requires_visible():
+    # https://github.com/conan-io/conan/issues/16058
+    # The "tool/0.1", even if visible doesn't generate any files in the "app" consumer side,
+    # even if the "app" declares it in build_context_activated, because it only works for direct
+    # dependencies. tool_requires(..., visible=True) only affects at the moment and version conflict
+    # detection
     c = TestClient()
     c.save({"tool/conanfile.py": GenConanfile("tool", "0.1").with_package_type("application"),
-            "pkgb/conanfile.py": GenConanfile("pkgb", "0.1").with_tool_requirement("tool/0.1", visible=True),
+            "pkgb/conanfile.py": GenConanfile("pkgb", "0.1").with_tool_requirement("tool/0.1",
+                                                                                   visible=True),
             "app/conanfile.py": GenConanfile("app", "0.1").with_requires("pkgb/0.1")
                                                           .with_settings("build_type")
-                                                          .with_tool_requirement("tool/0.1", visible=True)
                                                           .with_generator("CMakeDeps")})
 
     c.run("create tool")
     c.run("create pkgb")
     c.run("install app")
-    print(c.current_folder)
     cmake = c.load("app/pkgb-release-data.cmake")
+    # pkgb doesn't deepnd on tool, because it is a tool-require and not build-context activated
     assert "list(APPEND pkgb_FIND_DEPENDENCY_NAMES )" in cmake
-

--- a/conans/test/functional/toolchains/cmake/cmakedeps/test_build_context_transitive_build.py
+++ b/conans/test/functional/toolchains/cmake/cmakedeps/test_build_context_transitive_build.py
@@ -135,3 +135,21 @@ def test_error_cmakedeps_transitive_build_requires():
     c.run("export tool")
     c.run("install consumer --build=missing -s:b build_type=Release -s:h build_type=Debug")
     assert "tool/0.1: Created package" in c.out
+
+
+def test():
+    c = TestClient()
+    c.save({"tool/conanfile.py": GenConanfile("tool", "0.1").with_package_type("application"),
+            "pkgb/conanfile.py": GenConanfile("pkgb", "0.1").with_tool_requirement("tool/0.1", visible=True),
+            "app/conanfile.py": GenConanfile("app", "0.1").with_requires("pkgb/0.1")
+                                                          .with_settings("build_type")
+                                                          .with_tool_requirement("tool/0.1", visible=True)
+                                                          .with_generator("CMakeDeps")})
+
+    c.run("create tool")
+    c.run("create pkgb")
+    c.run("install app")
+    print(c.current_folder)
+    cmake = c.load("app/pkgb-release-data.cmake")
+    assert "list(APPEND pkgb_FIND_DEPENDENCY_NAMES )" in cmake
+

